cpg-analysis/src/main/kotlin/de/fraunhofer/aisec/cpg/query/QueryStates.kt

```diff
@@ -0,0 +1,236 @@
+ * limitations under the License.
+ *
+ *                    $$$$$$\  $$$$$$$\   $$$$$$\
+ *                   $$  __$$\ $$  __$$\ $$  __$$\
+ *                   $$ /  \__|$$ |  $$ |$$ /  \__|
+ *                   $$ |      $$$$$$$  |$$ |$$$$\
+ *                   $$ |      $$  ____/ $$ |\_$$ |
+ *                   $$ |  $$\ $$ |      $$ |  $$ |
+ *                   \$$$$$   |$$ |      \$$$$$   |
+ *                    \______/ \__|       \______/
+ *
+ */
+package de.fraunhofer.aisec.cpg.query
+
+import de.fraunhofer.aisec.cpg.assumptions.AssumptionStatus
+import de.fraunhofer.aisec.cpg.evaluation.compareTo
+
+sealed class RequirementState
+
+/**
+ * Represents a query that has been evaluated and the result is known to be `false` or some
+ * assumptions are rejected.
+ */
+data object Failed : RequirementState()
+
+/**
+ * Represents a query that has been evaluated and the result is known to be `true` and all
+ * assumptions are accepted.
+ */
+data object Succeeded : RequirementState()
+
+/**
+ * Represents a query that has been evaluated but the result is not yet known because some
+ * assumptions have to be accepted (or rejected).
+ */
+data object IsUndecided : RequirementState()
+
+/**
+ * Represents a query that has not yet been evaluated. Will be most likely used in the context of
+ * manual assessments which have to be conducted.
+ */
+data object IsNotYetEvaluated : RequirementState()
+
+fun QueryTree<Boolean>.toQueryState(): QueryTree<RequirementState> {
+    val (newValue, stringInfo) =
+        when {
+            !this.value || this.assumptions.any { it.status == AssumptionStatus.Rejected } ->
+                Failed to
+                    (if (!this.value) "The query was evaluated to false"
+                    else
+                        "The assumptions ${this.assumptions.filter { it.status == AssumptionStatus.Rejected }.map { it.id.toHexDashString() } } were rejected")
+            this.assumptions.any { it.status == AssumptionStatus.Undecided } ->
+                IsUndecided to
+                    "The assumptions ${this.assumptions.filter { it.status == AssumptionStatus.Undecided }.map { it.id.toHexDashString() }} are not yet decided"
+            this.value ==
+                this.assumptions.all {
+                    it.status == AssumptionStatus.Ignored || it.status == AssumptionStatus.Accepted
+                } ->
+                Succeeded to "The query was evaluated to true and all assumptions were accepted."
+            else -> IsNotYetEvaluated to "Something went wrong"
+        }
+
+    return QueryTree(
+        value = newValue,
+        children = mutableListOf(this),
+        stringRepresentation = "The requirement ${ newValue::class.simpleName } because $stringInfo",
+    )
+}
+
+/**
+ * Performs a logical and (&&) operation between the values and creates and returns [QueryTree]s.
+ */
+infix fun RequirementState.and(other: RequirementState): QueryTree<RequirementState> {
```

Comment:
I am confused. Why are all of these functions below necessary? Shouldn't it behave just like any other query tree?

Reply:
Because these functions are only available for `QueryTree<Boolean>`, the rest don't have that. So, if you want some logic (which is going to be the case based on the anticipated use-case) , you need this

Reply:
They way I am imagining it is the following: user-facing we will try to work with QueryTree<Boolean> and it the logic-operators should only be there for Boolean. Then once we execute the queries inside codyze, we convert the Boolean result + assumptions into the QueryTree<RequirementState>. At this state, we will only have the (final) result. We will not do any more computations with it.